compilation of xgboost failed with an error

compilation of xgboost failed with the following error 
```
[root@ build]# cmake ../
-- The CXX compiler identification is GNU 8.3.1
-- The C compiler identification is GNU 8.3.1
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- CMake version 3.13.0
-- Performing Test XGBOOST_MM_PREFETCH_PRESENT
-- Performing Test XGBOOST_MM_PREFETCH_PRESENT - Failed
-- Performing Test XGBOOST_BUILTIN_PREFETCH_PRESENT
-- Performing Test XGBOOST_BUILTIN_PREFETCH_PRESENT - Success
-- xgboost VERSION: 1.1.0
-- Setting build type to 'Release' as none was specified.
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Found OpenMP_C: -fopenmp (found version "4.5")
-- Found OpenMP_CXX: -fopenmp (found version "4.5")
-- Found OpenMP: TRUE (found version "4.5")
-- Found OpenMP_C: -fopenmp (found version "4.5")
-- Found OpenMP_CXX: -fopenmp (found version "4.5")
-- Looking for clock_gettime in rt
-- Looking for clock_gettime in rt - found
-- Looking for fopen64
-- Looking for fopen64 - not found
-- Looking for C++ include cxxabi.h
-- Looking for C++ include cxxabi.h - found
-- Looking for nanosleep
-- Looking for nanosleep - found
-- Looking for backtrace
-- Looking for backtrace - found
-- backtrace facility detected in default set of libraries
-- Found Backtrace: /usr/include
-- Check if the system is big endian
-- Searching 16 bit integer
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of unsigned short
-- Check size of unsigned short - done
-- Using unsigned short
-- Check if the system is big endian - big endian
-- /root/xgboost/dmlc-core/cmake/build_config.h.in -> include/dmlc/build_config.h
-- Performing Test SUPPORT_MSSE2
-- Performing Test SUPPORT_MSSE2 - Failed
DMLC_ROOT point to /root/xgboost/rabit/../dmlc-core
-- Configuring done
-- Generating done
-- Build files have been written to: /root/xgboost/build
[root@tureens1 build]# make -j$(nproc)
Scanning dependencies of target dmlc
[  1%] Building CXX object dmlc-core/CMakeFiles/dmlc.dir/src/config.cc.o
[  2%] Building CXX object dmlc-core/CMakeFiles/dmlc.dir/src/io.cc.o
[  3%] Building CXX object dmlc-core/CMakeFiles/dmlc.dir/src/io/line_split.cc.o
[  6%] Building CXX object dmlc-core/CMakeFiles/dmlc.dir/src/recordio.cc.o
[  6%] Building CXX object dmlc-core/CMakeFiles/dmlc.dir/src/io/recordio_split.cc.o
[  6%] Building CXX object dmlc-core/CMakeFiles/dmlc.dir/src/data.cc.o
[  7%] Building CXX object dmlc-core/CMakeFiles/dmlc.dir/src/io/indexed_recordio_split.cc.o
[  9%] Building CXX object dmlc-core/CMakeFiles/dmlc.dir/src/io/input_split_base.cc.o
[ 10%] Building CXX object dmlc-core/CMakeFiles/dmlc.dir/src/io/filesys.cc.o
[ 11%] Building CXX object dmlc-core/CMakeFiles/dmlc.dir/src/io/local_filesys.cc.o
[ 12%] Linking CXX static library libdmlc.a
[ 12%] Built target dmlc
Scanning dependencies of target rabit_base
Scanning dependencies of target objxgboost
Scanning dependencies of target rabit_mock
Scanning dependencies of target rabit_empty
Scanning dependencies of target rabit_mock_static
Scanning dependencies of target rabit
[ 13%] Building CXX object rabit/CMakeFiles/rabit_empty.dir/src/engine_empty.cc.o
[ 14%] Building CXX object rabit/CMakeFiles/rabit_empty.dir/src/c_api.cc.o
[ 15%] Building CXX object rabit/CMakeFiles/rabit_base.dir/src/allreduce_base.cc.o
[ 17%] Building CXX object rabit/CMakeFiles/rabit_base.dir/src/engine_base.cc.o
[ 18%] Building CXX object rabit/CMakeFiles/rabit_mock.dir/src/allreduce_base.cc.o
[ 19%] Building CXX object rabit/CMakeFiles/rabit.dir/src/allreduce_base.cc.o
[ 20%] Building CXX object rabit/CMakeFiles/rabit_mock_static.dir/src/allreduce_base.cc.o
[ 21%] Building CXX object src/CMakeFiles/objxgboost.dir/c_api/c_api.cc.o
[ 22%] Building CXX object src/CMakeFiles/objxgboost.dir/c_api/c_api_error.cc.o
[ 23%] Building CXX object src/CMakeFiles/objxgboost.dir/common/common.cc.o
[ 25%] Linking CXX static library librabit_empty.a
[ 25%] Built target rabit_empty
[ 26%] Building CXX object src/CMakeFiles/objxgboost.dir/common/hist_util.cc.o
[ 27%] Building CXX object rabit/CMakeFiles/rabit_base.dir/src/c_api.cc.o
[ 28%] Building CXX object rabit/CMakeFiles/rabit.dir/src/allreduce_robust.cc.o
[ 29%] Building CXX object src/CMakeFiles/objxgboost.dir/common/host_device_vector.cc.o
[ 31%] Building CXX object rabit/CMakeFiles/rabit_mock.dir/src/allreduce_robust.cc.o
[ 31%] Building CXX object rabit/CMakeFiles/rabit_mock_static.dir/src/allreduce_robust.cc.o
[ 32%] Linking CXX static library librabit_base.a
[ 34%] Building CXX object rabit/CMakeFiles/rabit.dir/src/engine.cc.o
[ 34%] Built target rabit_base
[ 35%] Building CXX object rabit/CMakeFiles/rabit_mock.dir/src/engine_mock.cc.o
[ 36%] Building CXX object rabit/CMakeFiles/rabit.dir/src/c_api.cc.o
[ 37%] Building CXX object rabit/CMakeFiles/rabit_mock.dir/src/c_api.cc.o
[ 38%] Building CXX object src/CMakeFiles/objxgboost.dir/common/io.cc.o
[ 39%] Building CXX object src/CMakeFiles/objxgboost.dir/common/json.cc.o
[ 40%] Building CXX object src/CMakeFiles/objxgboost.dir/common/probability_distribution.cc.o
[ 42%] Linking CXX static library librabit.a
[ 42%] Built target rabit
[ 43%] Building CXX object src/CMakeFiles/objxgboost.dir/common/survival_util.cc.o
[ 44%] Building CXX object rabit/CMakeFiles/rabit_mock_static.dir/src/engine_mock.cc.o
[ 45%] Linking CXX shared library librabit_mock.so
[ 45%] Built target rabit_mock
[ 46%] Building CXX object src/CMakeFiles/objxgboost.dir/common/timer.cc.o
[ 47%] Building CXX object src/CMakeFiles/objxgboost.dir/common/version.cc.o
[ 48%] Building CXX object src/CMakeFiles/objxgboost.dir/data/data.cc.o
[ 50%] Building CXX object src/CMakeFiles/objxgboost.dir/data/ellpack_page.cc.o
[ 51%] Building CXX object rabit/CMakeFiles/rabit_mock_static.dir/src/c_api.cc.o
[ 52%] Building CXX object src/CMakeFiles/objxgboost.dir/data/ellpack_page_source.cc.o
[ 53%] Building CXX object src/CMakeFiles/objxgboost.dir/data/simple_dmatrix.cc.o
[ 54%] Building CXX object src/CMakeFiles/objxgboost.dir/data/sparse_page_dmatrix.cc.o
[ 55%] Linking CXX static library librabit_mock_static.a
[ 55%] Built target rabit_mock_static
[ 56%] Building CXX object src/CMakeFiles/objxgboost.dir/data/sparse_page_raw_format.cc.o
In file included from /root/xgboost/dmlc-core/include/dmlc/io.h:446,
                 from /root/xgboost/src/data/data.cc:8:
/root/xgboost/dmlc-core/include/dmlc/./serializer.h: In instantiation of 'static void dmlc::serializer::IfThenElse<false, Then, Else, T>::Write(dmlc::Stream*, const T&) [with Then = dmlc::serializer::SaveLoadClassHandler<xgboost::DataType>; Else = dmlc::serializer::UndefinedSerializerFor<xgboost::DataType>; T = xgboost::DataType]':
/root/xgboost/dmlc-core/include/dmlc/./serializer.h:63:16:   recursively required from 'static void dmlc::serializer::IfThenElse<false, Then, Else, T>::Write(dmlc::Stream*, const T&) [with Then = dmlc::serializer::NativePODHandler<xgboost::DataType>; Else = dmlc::serializer::IfThenElse<false, dmlc::serializer::SaveLoadClassHandler<xgboost::DataType>, dmlc::serializer::UndefinedSerializerFor<xgboost::DataType>, xgboost::DataType>; T = xgboost::DataType]'
/root/xgboost/dmlc-core/include/dmlc/./serializer.h:63:16:   required from 'static void dmlc::serializer::IfThenElse<false, Then, Else, T>::Write(dmlc::Stream*, const T&) [with Then = dmlc::serializer::ArithmeticHandler<xgboost::DataType>; Else = dmlc::serializer::IfThenElse<false, dmlc::serializer::NativePODHandler<xgboost::DataType>, dmlc::serializer::IfThenElse<false, dmlc::serializer::SaveLoadClassHandler<xgboost::DataType>, dmlc::serializer::UndefinedSerializerFor<xgboost::DataType>, xgboost::DataType>, xgboost::DataType>; T = xgboost::DataType]'
/root/xgboost/dmlc-core/include/dmlc/./serializer.h:275:16:   required from 'static void dmlc::serializer::Handler<T>::Write(dmlc::Stream*, const T&) [with T = xgboost::DataType]'
/root/xgboost/dmlc-core/include/dmlc/io.h:452:32:   required from 'void dmlc::Stream::Write(const T&) [with T = xgboost::DataType]'
/root/xgboost/src/data/data.cc:40:14:   required from 'void {anonymous}::SaveScalarField(dmlc::Stream*, const string&, xgboost::DataType, const T&) [with T = long unsigned int; std::__cxx11::string = std::__cxx11::basic_string<char>]'
/root/xgboost/src/data/data.cc:158:63:   required from here
/root/xgboost/dmlc-core/include/dmlc/./serializer.h:63:16: error: 'Write' is not a member of 'dmlc::serializer::UndefinedSerializerFor<xgboost::DataType>'
     Else::Write(strm, data);
     ~~~~~~~~~~~^~~~~~~~~~~~
/root/xgboost/dmlc-core/include/dmlc/./serializer.h: In instantiation of 'static bool dmlc::serializer::IfThenElse<false, Then, Else, T>::Read(dmlc::Stream*, T*) [with Then = dmlc::serializer::SaveLoadClassHandler<xgboost::DataType>; Else = dmlc::serializer::UndefinedSerializerFor<xgboost::DataType>; T = xgboost::DataType]':
/root/xgboost/dmlc-core/include/dmlc/./serializer.h:66:22:   recursively required from 'static bool dmlc::serializer::IfThenElse<false, Then, Else, T>::Read(dmlc::Stream*, T*) [with Then = dmlc::serializer::NativePODHandler<xgboost::DataType>; Else = dmlc::serializer::IfThenElse<false, dmlc::serializer::SaveLoadClassHandler<xgboost::DataType>, dmlc::serializer::UndefinedSerializerFor<xgboost::DataType>, xgboost::DataType>; T = xgboost::DataType]'
/root/xgboost/dmlc-core/include/dmlc/./serializer.h:66:22:   required from 'static bool dmlc::serializer::IfThenElse<false, Then, Else, T>::Read(dmlc::Stream*, T*) [with Then = dmlc::serializer::ArithmeticHandler<xgboost::DataType>; Else = dmlc::serializer::IfThenElse<false, dmlc::serializer::NativePODHandler<xgboost::DataType>, dmlc::serializer::IfThenElse<false, dmlc::serializer::SaveLoadClassHandler<xgboost::DataType>, dmlc::serializer::UndefinedSerializerFor<xgboost::DataType>, xgboost::DataType>, xgboost::DataType>; T = xgboost::DataType]'
/root/xgboost/dmlc-core/include/dmlc/./serializer.h:294:11:   required from 'static bool dmlc::serializer::Handler<T>::Read(dmlc::Stream*, T*) [with T = xgboost::DataType]'
/root/xgboost/dmlc-core/include/dmlc/io.h:456:38:   required from 'bool dmlc::Stream::Read(T*) [with T = xgboost::DataType]'
/root/xgboost/src/data/data.cc:74:3:   required from 'void {anonymous}::LoadScalarField(dmlc::Stream*, const string&, xgboost::DataType, T*) [with T = long unsigned int; std::__cxx11::string = std::__cxx11::basic_string<char>]'
/root/xgboost/src/data/data.cc:197:64:   required from here
/root/xgboost/dmlc-core/include/dmlc/./serializer.h:66:22: error: 'Read' is not a member of 'dmlc::serializer::UndefinedSerializerFor<xgboost::DataType>'
     return Else::Read(strm, data);
            ~~~~~~~~~~^~~~~~~~~~~~
[ 57%] Building CXX object src/CMakeFiles/objxgboost.dir/gbm/gblinear.cc.o
[ 59%] Building CXX object src/CMakeFiles/objxgboost.dir/gbm/gblinear_model.cc.o
[ 60%] Building CXX object src/CMakeFiles/objxgboost.dir/gbm/gbm.cc.o
In file included from /root/xgboost/dmlc-core/include/dmlc/./io.h:446,
                 from /root/xgboost/dmlc-core/include/dmlc/data.h:14,
                 from /root/xgboost/include/xgboost/data.h:11,
                 from /root/xgboost/src/data/./simple_dmatrix.h:11,
                 from /root/xgboost/src/data/simple_dmatrix.cc:7:
/root/xgboost/dmlc-core/include/dmlc/././serializer.h: In instantiation of 'static bool dmlc::serializer::IfThenElse<false, Then, Else, T>::Read(dmlc::Stream*, T*) [with Then = dmlc::serializer::SaveLoadClassHandler<xgboost::Entry>; Else = dmlc::serializer::UndefinedSerializerFor<xgboost::Entry>; T = xgboost::Entry]':
/root/xgboost/dmlc-core/include/dmlc/././serializer.h:66:22:   recursively required from 'static bool dmlc::serializer::IfThenElse<false, Then, Else, T>::Read(dmlc::Stream*, T*) [with Then = dmlc::serializer::NativePODHandler<xgboost::Entry>; Else = dmlc::serializer::IfThenElse<false, dmlc::serializer::SaveLoadClassHandler<xgboost::Entry>, dmlc::serializer::UndefinedSerializerFor<xgboost::Entry>, xgboost::Entry>; T = xgboost::Entry]'
/root/xgboost/dmlc-core/include/dmlc/././serializer.h:66:22:   required from 'static bool dmlc::serializer::IfThenElse<false, Then, Else, T>::Read(dmlc::Stream*, T*) [with Then = dmlc::serializer::ArithmeticHandler<xgboost::Entry>; Else = dmlc::serializer::IfThenElse<false, dmlc::serializer::NativePODHandler<xgboost::Entry>, dmlc::serializer::IfThenElse<false, dmlc::serializer::SaveLoadClassHandler<xgboost::Entry>, dmlc::serializer::UndefinedSerializerFor<xgboost::Entry>, xgboost::Entry>, xgboost::Entry>; T = xgboost::Entry]'
/root/xgboost/dmlc-core/include/dmlc/././serializer.h:294:11:   required from 'static bool dmlc::serializer::Handler<T>::Read(dmlc::Stream*, T*) [with T = xgboost::Entry]'
/root/xgboost/dmlc-core/include/dmlc/./io.h:456:38:   required from 'bool dmlc::Stream::Read(T*) [with T = xgboost::Entry]'
/root/xgboost/dmlc-core/include/dmlc/./io.h:469:9:   required from 'bool dmlc::Stream::ReadArray(T*, size_t) [with T = xgboost::Entry; size_t = long unsigned int]'
/root/xgboost/dmlc-core/include/dmlc/././serializer.h:165:58:   required from 'static bool dmlc::serializer::ComposeVectorHandler<T>::Read(dmlc::Stream*, std::vector<T>*) [with T = xgboost::Entry]'
/root/xgboost/dmlc-core/include/dmlc/././serializer.h:66:22:   required from 'static bool dmlc::serializer::IfThenElse<false, Then, Else, T>::Read(dmlc::Stream*, T*) [with Then = dmlc::serializer::NativePODVectorHandler<xgboost::Entry>; Else = dmlc::serializer::ComposeVectorHandler<xgboost::Entry>; T = std::vector<xgboost::Entry, std::allocator<xgboost::Entry> >]'
/root/xgboost/dmlc-core/include/dmlc/././serializer.h:312:11:   required from 'static bool dmlc::serializer::Handler<std::vector<T> >::Read(dmlc::Stream*, std::vector<T>*) [with T = xgboost::Entry]'
/root/xgboost/dmlc-core/include/dmlc/./io.h:456:38:   required from 'bool dmlc::Stream::Read(T*) [with T = std::vector<xgboost::Entry, std::allocator<xgboost::Entry> >]'
/root/xgboost/src/data/simple_dmatrix.cc:176:50:   required from here
/root/xgboost/dmlc-core/include/dmlc/././serializer.h:66:22: error: 'Read' is not a member of 'dmlc::serializer::UndefinedSerializerFor<xgboost::Entry>'
     return Else::Read(strm, data);
            ~~~~~~~~~~^~~~~~~~~~~~
/root/xgboost/dmlc-core/include/dmlc/././serializer.h: In instantiation of 'static void dmlc::serializer::IfThenElse<false, Then, Else, T>::Write(dmlc::Stream*, const T&) [with Then = dmlc::serializer::SaveLoadClassHandler<xgboost::Entry>; Else = dmlc::serializer::UndefinedSerializerFor<xgboost::Entry>; T = xgboost::Entry]':
/root/xgboost/dmlc-core/include/dmlc/././serializer.h:63:16:   recursively required from 'static void dmlc::serializer::IfThenElse<false, Then, Else, T>::Write(dmlc::Stream*, const T&) [with Then = dmlc::serializer::NativePODHandler<xgboost::Entry>; Else = dmlc::serializer::IfThenElse<false, dmlc::serializer::SaveLoadClassHandler<xgboost::Entry>, dmlc::serializer::UndefinedSerializerFor<xgboost::Entry>, xgboost::Entry>; T = xgboost::Entry]'
/root/xgboost/dmlc-core/include/dmlc/././serializer.h:63:16:   required from 'static void dmlc::serializer::IfThenElse<false, Then, Else, T>::Write(dmlc::Stream*, const T&) [with Then = dmlc::serializer::ArithmeticHandler<xgboost::Entry>; Else = dmlc::serializer::IfThenElse<false, dmlc::serializer::NativePODHandler<xgboost::Entry>, dmlc::serializer::IfThenElse<false, dmlc::serializer::SaveLoadClassHandler<xgboost::Entry>, dmlc::serializer::UndefinedSerializerFor<xgboost::Entry>, xgboost::Entry>, xgboost::Entry>; T = xgboost::Entry]'
/root/xgboost/dmlc-core/include/dmlc/././serializer.h:275:16:   required from 'static void dmlc::serializer::Handler<T>::Write(dmlc::Stream*, const T&) [with T = xgboost::Entry]'
/root/xgboost/dmlc-core/include/dmlc/./io.h:452:32:   required from 'void dmlc::Stream::Write(const T&) [with T = xgboost::Entry]'
/root/xgboost/dmlc-core/include/dmlc/./io.h:462:5:   required from 'void dmlc::Stream::WriteArray(const T*, size_t) [with T = xgboost::Entry; size_t = long unsigned int]'
/root/xgboost/dmlc-core/include/dmlc/././serializer.h:158:5:   required from 'static void dmlc::serializer::ComposeVectorHandler<T>::Write(dmlc::Stream*, const std::vector<T>&) [with T = xgboost::Entry]'
/root/xgboost/dmlc-core/include/dmlc/././serializer.h:63:16:   required from 'static void dmlc::serializer::IfThenElse<false, Then, Else, T>::Write(dmlc::Stream*, const T&) [with Then = dmlc::serializer::NativePODVectorHandler<xgboost::Entry>; Else = dmlc::serializer::ComposeVectorHandler<xgboost::Entry>; T = std::vector<xgboost::Entry, std::allocator<xgboost::Entry> >]'
/root/xgboost/dmlc-core/include/dmlc/././serializer.h:305:12:   required from 'static void dmlc::serializer::Handler<std::vector<T> >::Write(dmlc::Stream*, const std::vector<T>&) [with T = xgboost::Entry]'
/root/xgboost/dmlc-core/include/dmlc/./io.h:452:32:   required from 'void dmlc::Stream::Write(const T&) [with T = std::vector<xgboost::Entry, std::allocator<xgboost::Entry> >]'
/root/xgboost/src/data/simple_dmatrix.cc:185:45:   required from here
/root/xgboost/dmlc-core/include/dmlc/././serializer.h:63:16: error: 'Write' is not a member of 'dmlc::serializer::UndefinedSerializerFor<xgboost::Entry>'
     Else::Write(strm, data);
     ~~~~~~~~~~~^~~~~~~~~~~~
make[2]: *** [src/CMakeFiles/objxgboost.dir/build.make:206: src/CMakeFiles/objxgboost.dir/data/data.cc.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [src/CMakeFiles/objxgboost.dir/build.make:245: src/CMakeFiles/objxgboost.dir/data/simple_dmatrix.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:487: src/CMakeFiles/objxgboost.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
```

--- a/include/xgboost/data.h
+++ b/include/xgboost/data.h
@@ -566,8 +566,7 @@ struct Handler<xgboost::Entry> {
   }
 
   inline static bool Read(Stream* strm, xgboost::Entry* data) {
-    strm->Read(&data->index);
-    strm->Read(&data->fvalue);
+    return strm->Read(&data->index) && strm->Read(&data->fvalue);
   }
 };
 

--- a/include/xgboost/data.h
+++ b/include/xgboost/data.h
@@ -9,6 +9,7 @@
 
 #include <dmlc/base.h>
 #include <dmlc/data.h>
+#include <dmlc/serializer.h>
 #include <rabit/rabit.h>
 #include <xgboost/base.h>
 #include <xgboost/span.h>
@@ -554,5 +555,22 @@ inline BatchSet<EllpackPage> DMatrix::GetBatches(const BatchParam& param) {
 
 namespace dmlc {
 DMLC_DECLARE_TRAITS(is_pod, xgboost::Entry, true);
-}
+
+namespace serializer {
+
+template <>
+struct Handler<xgboost::Entry> {
+  inline static void Write(Stream* strm, const xgboost::Entry& data) {
+    strm->Write(data.index);
+    strm->Write(data.fvalue);
+  }
+
+  inline static bool Read(Stream* strm, xgboost::Entry* data) {
+    strm->Read(&data->index);
+    strm->Read(&data->fvalue);
+  }
+};
+
+}  // namespace serializer
+}  // namespace dmlc
 #endif  // XGBOOST_DATA_H_

--- a/src/data/data.cc
+++ b/src/data/data.cc
@@ -37,7 +37,7 @@ template <typename T>
 void SaveScalarField(dmlc::Stream *strm, const std::string &name,
                      xgboost::DataType type, const T &field) {
   strm->Write(name);
-  strm->Write(type);
+  strm->Write(static_cast<uint8_t>(type));
   strm->Write(true);  // is_scalar=True
   strm->Write(field);
 }
@@ -47,7 +47,7 @@ void SaveVectorField(dmlc::Stream *strm, const std::string &name,
                      xgboost::DataType type, std::pair<uint64_t, uint64_t> shape,
                      const std::vector<T>& field) {
   strm->Write(name);
-  strm->Write(type);
+  strm->Write(static_cast<uint8_t>(type));
   strm->Write(false);  // is_scalar=False
   strm->Write(shape.first);
   strm->Write(shape.second);
@@ -71,7 +71,9 @@ void LoadScalarField(dmlc::Stream* strm, const std::string& expected_name,
   CHECK(strm->Read(&name)) << invalid;
   CHECK_EQ(name, expected_name)
       << invalid << " Expected field: " << expected_name << ", got: " << name;
-  CHECK(strm->Read(&type)) << invalid;
+  uint8_t type_val;
+  CHECK(strm->Read(&type_val)) << invalid;
+  type = static_cast<xgboost::DataType>(type_val);
   CHECK(type == expected_type)
       << invalid << "Expected field of type: " << static_cast<int>(expected_type) << ", "
       << "got field type: " << static_cast<int>(type);
@@ -91,7 +93,9 @@ void LoadVectorField(dmlc::Stream* strm, const std::string& expected_name,
   CHECK(strm->Read(&name)) << invalid;
   CHECK_EQ(name, expected_name)
     << invalid << " Expected field: " << expected_name << ", got: " << name;
-  CHECK(strm->Read(&type)) << invalid;
+  uint8_t type_val;
+  CHECK(strm->Read(&type_val)) << invalid;
+  type = static_cast<xgboost::DataType>(type_val);
   CHECK(type == expected_type)
     << invalid << "Expected field of type: " << static_cast<int>(expected_type) << ", "
     << "got field type: " << static_cast<int>(type);

